File: issuer-list-ui/src/main/java/com/adyen/checkout/issuerlist/IssuerListSpinnerView.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 26/4/2019.
 */

package com.adyen.checkout.issuerlist;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatSpinner;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;

import com.adyen.checkout.base.PaymentComponentState;
import com.adyen.checkout.base.api.ImageLoader;
import com.adyen.checkout.base.model.payments.request.IssuerListPaymentMethod;
import com.adyen.checkout.base.ui.view.AdyenLinearLayout;
import com.adyen.checkout.core.log.LogUtil;
import com.adyen.checkout.core.log.Logger;
import com.adyen.checkout.issuerlist.ui.R;

import java.util.Collections;
import java.util.List;

public abstract class IssuerListSpinnerView
        <IssuerListPaymentMethodT extends IssuerListPaymentMethod, IssuerListComponentT extends IssuerListComponent<IssuerListPaymentMethodT>>
        extends AdyenLinearLayout<IssuerListOutputData, IssuerListConfiguration, PaymentComponentState, IssuerListComponentT>
        implements AdapterView.OnItemSelectedListener {
    private static final String TAG = LogUtil.getTag();

    private AppCompatSpinner mIssuersSpinner;

    private final IssuerListInputData mIdealInputData = new IssuerListInputData();

    private IssuerListSpinnerAdapter mIssuersAdapter;

    public IssuerListSpinnerView(@NonNull Context context) {
        this(context, null);
    }

    public IssuerListSpinnerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    // Regular View constructor
    @SuppressWarnings("JavadocMethod")
    public IssuerListSpinnerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutInflater.from(getContext()).inflate(R.layout.issuer_list_spinner_view, this, true);
    }

    @Override
    public void initView() {
        mIssuersSpinner = findViewById(R.id.spinner_issuers);
        mIssuersSpinner.setAdapter(mIssuersAdapter);
        mIssuersSpinner.setOnItemSelectedListener(this);
    }

    @SuppressWarnings("PMD.EmptyMethodInAbstractClassShouldBeAbstract")
    @Override
    protected void initLocalizedStrings(@NonNull Context localizedContext) {
        // no embedded localized strings on this view
    }

    @Override
    public void onComponentAttached() {
        mIssuersAdapter = new IssuerListSpinnerAdapter(getContext(),
                Collections.<IssuerModel>emptyList(),
                ImageLoader.getInstance(getContext(), getComponent().getConfiguration().getEnvironment()),
                getComponent().getPaymentMethod().getType(),
                hideIssuersLogo());
    }

    @Override
    protected void observeComponentChanges(@NonNull LifecycleOwner lifecycleOwner) {
        getComponent().getIssuersLiveData().observe(lifecycleOwner, createIssuersObserver());
    }

    @Override
    public boolean isConfirmationRequired() {
        return true;
    }

    @SuppressWarnings("PMD.EmptyMethodInAbstractClassShouldBeAbstract")
    @Override
    public void highlightValidationErrors() {
        // no implementation
    }

    public boolean hideIssuersLogo() {
        return false;
    }

    void onIssuersChanged(@Nullable List<IssuerModel> issuerList) {
        mIssuersAdapter.updateIssuers(issuerList);
    }

    @Override
    public void onItemSelected(@NonNull AdapterView<?> parent, @NonNull View view, int position, long id) {
        Logger.d(TAG, "onItemSelected - " + mIssuersAdapter.getItem(position).getName());
        mIdealInputData.setSelectedIssuer(mIssuersAdapter.getItem(position));
        getComponent().inputDataChanged(mIdealInputData);
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        mIssuersSpinner.setEnabled(enabled);
    }

    @SuppressWarnings("PMD.EmptyMethodInAbstractClassShouldBeAbstract")
    @Override
    public void onNothingSelected(@NonNull AdapterView<?> parent) {
        // nothing changed
    }

    private Observer<List<IssuerModel>> createIssuersObserver() {
        return new Observer<List<IssuerModel>>() {
            @Override
            public void onChanged(@Nullable List<IssuerModel> issuerList) {
                onIssuersChanged(issuerList);
            }
        };
    }

    @NonNull
    protected Observer<IssuerListOutputData> createOutputDataObserver() {
        return new Observer<IssuerListOutputData>() {
            @Override
            public void onChanged(@Nullable IssuerListOutputData idealOutputData) {
                // Component does not change selected IssuerModel, add validation UI here if that's needed
            }
        };
    }
}

File: issuer-list/src/main/java/com/adyen/checkout/issuerlist/internal/ui/view/IssuerListSpinnerView.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 26/4/2019.
 */
package com.adyen.checkout.issuerlist.internal.ui.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.AdapterView
import android.widget.LinearLayout
import com.adyen.checkout.components.core.internal.ui.ComponentDelegate
import com.adyen.checkout.core.old.AdyenLogLevel
import com.adyen.checkout.core.old.internal.util.adyenLog
import com.adyen.checkout.issuerlist.databinding.IssuerListSpinnerViewBinding
import com.adyen.checkout.issuerlist.internal.ui.IssuerListDelegate
import com.adyen.checkout.ui.core.internal.ui.ComponentView
import kotlinx.coroutines.CoroutineScope

internal class IssuerListSpinnerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) :
    LinearLayout(
        context,
        attrs,
        defStyleAttr,
    ),
    ComponentView,
    AdapterView.OnItemSelectedListener {

    private val binding: IssuerListSpinnerViewBinding =
        IssuerListSpinnerViewBinding.inflate(LayoutInflater.from(context), this)

    private lateinit var issuersAdapter: IssuerListSpinnerAdapter

    private lateinit var localizedContext: Context

    private lateinit var issuerListDelegate: IssuerListDelegate<*, *>

    override fun initView(delegate: ComponentDelegate, coroutineScope: CoroutineScope, localizedContext: Context) {
        require(delegate is IssuerListDelegate<*, *>) { "Unsupported delegate type" }
        issuerListDelegate = delegate

        this.localizedContext = localizedContext
        initLocalizedStrings(localizedContext)

        issuersAdapter = IssuerListSpinnerAdapter(
            context = context,
            issuerList = delegate.getIssuers(),
            paymentMethod = delegate.getPaymentMethodType(),
            hideIssuerLogo = delegate.componentParams.hideIssuerLogos,
        )
        binding.spinnerIssuers.apply {
            adapter = issuersAdapter
            onItemSelectedListener = this@IssuerListSpinnerView
        }
    }

    @Suppress("UnusedPrivateMember", "UNUSED_PARAMETER")
    private fun initLocalizedStrings(localizedContext: Context) {
        // no embedded localized strings on this view
    }

    override fun highlightValidationErrors() {
        // no implementation
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val item = issuersAdapter.getItem(position)
        adyenLog(AdyenLogLevel.DEBUG) { "onItemSelected - " + item.name }
        issuerListDelegate.updateInputData { selectedIssuer = item }
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        binding.spinnerIssuers.isEnabled = enabled
    }

    override fun onNothingSelected(parent: AdapterView<*>) {
        // nothing changed
    }

    override fun getView(): View = this
}
